client/components/tinymce/plugins/mentions/mentions.jsx

```diff
@@ -0,0 +1,281 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import ReactDomServer from 'react-dom/server';
+import { connect } from 'react-redux';
+import { findIndex, get, head, includes, last, pick } from 'lodash';
+import tinymce from 'tinymce/tinymce';
+
+/**
+ * Internal dependencies
+ */
+import SuggestionList from './suggestion-list';
+import EditorMention from './editor-mention';
+import QueryUsersSuggestions from 'components/data/query-users-suggestions';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getUserSuggestions } from 'state/users/suggestions/selectors';
+
+/**
+ * Module variables
+ */
+// TinyMCE mapping of common keycodes.
+const VK = tinymce.util.VK;
+
+export class Mentions extends Component {
+	matchingSuggestions = [];
+
+	state = {
+		popoverContext: null,
+		query: '',
+		showPopover: false,
+	};
+
+	componentDidMount() {
+		const { editor } = this.props;
+		const { left, top } = this.getPosition();
+
+		editor.on( 'keyup', this.onKeyUp );
+		editor.on( 'click', this.hidePopover );
+
+		this.left = left;
+		this.top = top;
+	}
+
+	componentWillUpdate( nextProps, nextState ) {
+		// Update position of popover if going from invisible to visible state.
+		if ( ! this.state.showPopover && nextState.showPopover ) {
+			this.updatePosition( nextState );
+			this.props.editor.on( 'keydown', this.onKeyDown );
+
+			return;
+		}
+
+		// Visible to invisible state.
+		if ( this.state.showPopover && ! nextState.showPopover ) {
+			this.props.editor.off( 'keydown', this.onKeyDown );
+
+			return;
+		}
+
+		// Update position of popover if cursor has moved to a new line.
+		if ( nextState.showPopover ) {
+			const { top, left } = this.getPosition();
+			const isLineBefore = ( this.top > top ) && ( this.left < left );
+			const isLineAfter = ( this.top < top ) && ( this.left > left );
+
+			if ( isLineBefore || isLineAfter ) {
+				this.updatePosition( nextState, { top, left } );
+			}
+		}
+	}
+
+	componentWillUnmount() {
+		const { editor } = this.props;
+
+		editor.off( 'keydown', this.onKeyDown );
+		editor.off( 'keyup', this.onKeyUp );
+		editor.off( 'click', this.hidePopover );
+	}
+
+	getMatchingSuggestions( suggestions, query ) {
+		if ( query ) {
+			suggestions = suggestions.filter( ( { user_login: userLogin, display_name: displayName } ) => {
+				// Start of string or preceded by a space.
+				const matcher = new RegExp( `^${ query }|\\s${ query }`, 'ig' );
+
+				return matcher.test( `${ userLogin } ${ displayName }` );
+			} );
+		}
+
+		return suggestions.slice( 0, 10 );
```

Comment:
I was just looking at [`getUserSuggestions()`](https://github.com/Automattic/wp-calypso/blob/e7330d483b54b57ec71588e21ef031ca499cf951/client/state/users/suggestions/selectors.js#L21) and I wonder if it would make sense to fall back to an empty array there?